decisions/websre/0004-use-1-kubernetes-cluster-per-environment.md

```diff
@@ -0,0 +1,56 @@
+# 4. Use 1 Kubernetes Cluster Per Environment
+
+Date: 2021-06-09
+Scope: Web SRE Team
+
+## Status
+
+Proposed
+
+## Context
+
+We need to have multiple environments (sometimes called tiers) for Web SRE Services. These environments generally serve different purposes & audiences, and can have different uptime requirements. Environments can optionally serve as required points in the path towards the production environment. 
+
+At Mozilla, our service environments generally include Development (Dev), Staging (Stage), and Production (Prod). There can sometimes also be a Testing environment (usually ephemeral). Generally, these environments are in the same infrastructure provider account, and differ at the infrastructure or application platform level (e.g. at the Kubernetes level) - with the exception of Testing, which tends to be ephemeral within a particular context (e.g. a docker-compose spin up or a kind cluster in a CI pipeline).
+
+There are multiple ways that we differentiate these persistent (dev, stage, prod) environments on our infrastructure and application platforms. Within a Kubernetes context, some common patterns for separating persistent environments are:
+1. Having a distinct Kubernetes cluster per environment;
+2. Having one namespace per environment in a shared Kubernetes cluster;
+3. Having distinct node pool(s) per environment within a single cluster along with using selectors, taints, and tolerations.
+
+## Decision
+
+For Web SRE Services persistent environments - dev, stage, prod - that are Kubernetes-based, we will:
+* Use a distinct Kubernetes cluster per environment (dev, stage, prod) within the same infrastructure account;
```

Comment:
Having 3 clusters seems like a lot.  Would we prefer non-prod/prod? I think that's more of a $$$/maintaining all the clusters decision vs a technical choice.  Either way would be ok with me, if money was no issue having 3 clusters sounds great.

Reply:
I tentatively support the dev/stage/prod model when we have active development, and I definitely support disabling dev when we do not. (I wouldn't expect to see a Refractr 'dev', just stage/prod, for example.)

Reply:
I'm on-board with calling the k8s namespaces dev/stage/prod no matter what.  I guess I was mostly thinking about the clusters themselves.  Moving from only having one cluster, to having two is a clear improvement, but I'm not sure that three clusters is better than two.  All the extra work in building and maintaining a cluster just to separate dev and staging may not be worth it.

  Having a cluster that groups dev/stage together and a separate one for prod seems reasonable.  dev/stage/prod as unique clusters for really high volume apps under a ton of active development also seems reasonable.  Not sure how to get the ADR to endorse either of those options.

Reply:
As one example, I'd love to have a dev cluster for refractr & similar apps, & could see the usage points for it right away - deployments of the application still go through stage & prod un-impacted by infrastructure work done in dev.

I also think we can get better with kubernetes cluster deploys so they are easier to handle - I mean, some gke deployment patterns i've worked with involved a whole new prod cluster for an app then a blue-green swap after smoke testing. once we got our infra patterns for the gke cluster setup, it was pretty straight forward to manage clusters.

if for whatever reason a service really wants to mix environments in clusters again, the SRE can always note they don't want to follow the standards and do what they want, as long as its documented. I think it'll come back to why this ADR came up to begin with, so maybe we just make sure production is protected and ask people to try to keep stage, dev separate unless there is a real reason in the moment not to.

Reply:
I don't think that two clusters should be an exception to our standard procedures and processes.  Rather than treat it like an exception to this ADR, I'd prefer the ADR to explicitly accept both the two and three cluster model.  I think two clusters is a really good compromise for where we are today in terms of infrastructure automation and the number of clusters/environments we have.  This shouldn't preclude moving to three clusters for apps that need or want it (like refractr).

Reply:
For me, this ADR is tricky because it defines how we map k8s to environments before we've built up a standard menu of environments, their purposes, the support we offer them, etc. The important quality I think we all agree on is that we need another environment, distinct from the production environment at the infrastructure level, for testing changes both to the application and to the infrastructure. Thus, having this ADR encode our default approach as having a stage and prod deployment of the app, each on a separate cluster, makes sense to me.

I'm not convinced we should go beyond that and dictate that other environments require separate clusters. My preference is to leave that decision to the discretion of the SREs working on an application based on factors like availability and security requirements. I think there may be cases where we want a dev deployment on a separate cluster, and cases where we decide it's okay to run on the same cluster as stage. Likewise, I can anticipate sometimes having multiple prod deployments, which may or may not make sense to run on the same cluster.